salt/states/module.py

```diff
@@ -314,22 +314,31 @@ def _call_function(name, returner=None, **kwargs):
     :return:
     '''
     argspec = salt.utils.args.get_function_argspec(__salt__[name])
+
+    # func_kw is initialized to a dictinary of keyword arguments the function to be run accepts
     func_kw = dict(zip(argspec.args[-len(argspec.defaults or []):],  # pylint: disable=incompatible-py3-code
                    argspec.defaults or []))
+
+    # func_args is initialized to a list of keyword arguments the function to be run accepts 
+    func_args = argspec.args[:len(argspec.args or [] ) - len(argspec.defaults or [])]
     arg_type, na_type, kw_type = [], {}, False
     for funcset in reversed(kwargs.get('func_args') or []):
         if not isinstance(funcset, dict):
-            kw_type = True
-        if kw_type:
-            if isinstance(funcset, dict):
-                arg_type += funcset.values()
-                na_type.update(funcset)
-            else:
-                arg_type.append(funcset)
+            # We are just receiving a list of args to the function to be run, so just append
+            # those to the arg list that we will pass to the func.
+            arg_type.append(funcset)
         else:
-            func_kw.update(funcset)
+            for kwarg_key in funcset.keys():
```

Comment:
Do we really have to call keys?

Reply:
No. I just was as explicit as I could be because this code is so dense and hard to read.

Reply:
But that does create a new list... :)